### PR TITLE
Scc 3938/bib details styling

### PR DESCRIPTION
**Pull request by charmingduchess**
Add styles to bib details. 
Most additions are overrides for DS List component.Are there any DS List props for not rendering the default html styling but leaving the DS CSS? noStyling seems to remove all styles including DS, but I just want to remove the numbering and indentation from the `<ol>` elements.

### CODE REVIEW

**Comment by EdwinGuzman** [comment]
> Are there any DS List props for not rendering the default html styling but leaving the DS CSS? noStyling seems to remove all styles including DS, but I just want to remove the numbering and indentation from the <ol> elements.

No, there isn't a prop for this. I believe `noStyling` was intended to work that way and then the consuming app would add overrides but this makes me think it should have an update. I'll take a closer look at this PR and for visibility, tagging @bigfishdesign13

File: src/components/BibPage/BibDetail.tsx [hunk 1]
```diff
@@ -0,0 +1,142 @@
...
+  details: BibDetail[] | LinkedBibDetail[]
+  heading?: string
+}
+
+const BibDetails = ({ details, heading }: BibDetailsProps) => {
+  return (
+    details?.length > 0 && (
+      <>
+        {heading && <Heading level="three">{heading}</Heading>}
+        <List
```
Comment (EdwinGuzman) [inline src/components/BibPage/BibDetail.tsx:26]: The `List` component allows a `title` prop that renders an h2 heading.

The latest DS version [2.1.4](https://www.npmjs.com/package/@nypl/design-system-react-components/v/2.1.4) allows for the title to be a custom component. So this can be:
```
<List
  title={heading && <Heading level="three">{heading}</Heading>}
  ...
```

Can you try this default `List` style for the heading and the row of dt/dds? This would mean removing the `sx={{ borderBottom: "none" }}` rule.

I'd like to see what Apporva/Marty think about this update on the bib page. If it doesn't work out and we want the older styles, that's fine but we should try to use the default DS styles.

File: src/components/BibPage/BibDetail.tsx [hunk 2]
```diff
@@ -0,0 +1,142 @@
...
+    )
+  )
+}
+
+const buildDetailElement = (field: BibDetail) => {
+  return (
+    <>
+      <dt>{field.label}</dt>
+      <dd>
+        <List data-testId={`${field.label}`} type="ol">
```
Comment (EdwinGuzman) [inline src/components/BibPage/BibDetail.tsx:56]: Regarding the question in the PR's description: why the use of `ol`? The current page uses `ul` and I'm not sure if the values for each field should be numbered.

File: styles/components/BibDetails.module.scss [hunk 3]
```diff
@@ -0,0 +1,14 @@
+@import "../utils/mixins";
+
+.bibDetails {
+  dl > dt,
+  dl > dd {
+    border: none;
```
Comment (EdwinGuzman) [inline styles/components/BibDetails.module.scss:6]: Can you leave the default border to see what Apoorva thinks?

File: src/components/BibPage/BibDetail.tsx (hunk 2 above)
Comment (charmingduchess) [reply src/components/BibPage/BibDetail.tsx]: I'm using ol because the order matters for the primary/parallel matches. but maybe i'm misinterpreting the use case for an ol. happy to use whatever!

File: styles/components/BibDetails.module.scss (hunk 3 above)
Comment (charmingduchess) [reply styles/components/BibDetails.module.scss:6]: I spoke with @apoorvakhanolkar earlier about this and she requested that it be removed.

Comment (charmingduchess) [reply styles/components/BibDetails.module.scss:6]: Here are screenshots of the two different styling options. @apoorvakhanolkar can you confirm which is preferable?
Screen Shot 2024-01-12 at 9 36 45 AM
Screen Shot 2024-01-12 at 9 36 35 AM

Comment (charmingduchess) [reply styles/components/BibDetails.module.scss:6]: I think the border looks classy 😎 . Looks much better now with the headings too, imo.

Comment (apoorvakhanolkar) [reply styles/components/BibDetails.module.scss:6]: I prefer it without the borders! Just looks a bit cleaner to me.

**Review by EdwinGuzman — APPROVED** [review]
Looks good! I'd wait on Diego's review but I'm curious, would the order of merge create conflicts with this and the Bib page route PR?

File: src/models/BibDetails.ts [hunk 4]
```diff
@@ -0,0 +1,209 @@
...
+        let detail
+        if (fieldMapping.field === "supplementaryContent")
+          detail = this.supplementaryContent
+        else if (fieldMapping.field === "creatorLiteral")
+          detail = this.buildInternalLinkedDetail(fieldMapping)
+        else detail = this.buildStandardDetail(fieldMapping)
+        if (!detail) return
+        return detail
+      })
+      .filter((f) => f)
```
Comment (dgcohen) [inline src/models/BibDetails.ts:114]: Out of curiosity, what are these filters doing?

File: src/utils/bibModelPreprocessing.ts [hunk 5]
```diff
@@ -0,0 +1,110 @@
...
+
+/**
+ * interleaveParallel(arr1, arr2)
+ * Given two arrays, returns the elements interleaved, with falsey elements removed.
+ * Also combines data from matching elements when necessary.
+ * Example: interleaveParallel ([1, 2, null, 3], [5,6,7,8,9]) =>
+ * [1,5,2,6,7,3,8,9].
+ * Assumes that arr2 is at least as long as arr1.
+ *
+ * @param {array} arr1
```
Comment (dgcohen) [inline src/utils/bibModelPreprocessing.ts:63]: Don't think we need to be using JS doc type annotations for this project since we're using Typescript, but let me know if you disagree.

File: src/models/BibDetails.ts (hunk 4 above)
Comment (charmingduchess) [reply src/models/BibDetails.ts:114]: filtering out null values. If a bib doesn't have a supplementary content, for example, we don't want there to be a null value in the array of details.

File: src/utils/bibModelPreprocessing.ts (hunk 5 above)
Comment (charmingduchess) [reply src/utils/bibModelPreprocessing.ts:63]: True! this was just an artifact from DFE.

**Comment by charmingduchess** [comment]
@EdwinGuzman  it's possible. is there part of the code you're concerned about in particular? i did create and then rename a `pages/bib.tsx` file, but i don't think it'll be too hard to parse out if there's a conflict.

**Comment by EdwinGuzman** [comment]
Nothing in particular that I'm concerned about, but it would be good to consider -- someone may end up with merge conflicts and that's okay but hope nothing gets overridden accidentally.

File: src/components/BibPage/BibDetail.tsx [hunk 6]
```diff
@@ -0,0 +1,129 @@
...
+              return buildPlainTextElement(detail as BibDetail)
+            }
+          }
+        )}
+      </List>
+    )
+  )
+}
+
+const buildDetailElement = (label: string, listChildren: ReactNode[]) => {
```
Comment (dgcohen) [inline src/components/BibPage/BibDetail.tsx:51]: Small nit, but what do you think of renaming these to DetailElement, PlainTextElement and CompoundSubjectHeadingElement? I know they're technically functions but I'd rather keep the naming consistent with other react elements.

File: src/components/BibPage/BibDetail.tsx [hunk 7]
```diff
@@ -0,0 +1,129 @@
...
+      </dd>
+    </>
+  )
+}
+
+const buildPlainTextElement = (field: BibDetail) => {
+  const values = field.value.map((val: string, i: number) => {
+    const stringDirection = displayRtl(val)
+    return (
+      <li dir={stringDirection} key={i}>
```
Comment (dgcohen) [inline src/components/BibPage/BibDetail.tsx:68]: This key might need a prefix to avoid duplicates.

File: src/components/BibPage/BibDetail.tsx [hunk 8]
```diff
@@ -0,0 +1,129 @@
...
+    </li>
+  ))
+  return buildDetailElement(field.label, values)
+}
+
+const buildSingleSubjectHeadingElement = (subjectHeadingUrls: Url[]) => {
+  const urls = subjectHeadingUrls.reduce((linksPerSubject, url: Url, index) => {
+    const divider = (
+      // this span will render as > in between the divided subject heading links
+      <span data-testid="divider" key={`divider-${index}`}>
```
Comment (dgcohen) [inline src/components/BibPage/BibDetail.tsx:94]: Can this be achieved with CSS (pseudoselector with a content attribute?). Not sure if it can but curious.

File: src/components/BibPage/BibDetail.tsx [hunk 9]
```diff
@@ -0,0 +1,129 @@
...
+    }
+    return linksPerSubject
+  }, [] as React.JSX.Element[])
+  return urls
+}
+
+const buildLinkedElement = (field: LinkedBibDetail) => {
+  const internalOrExternal = field.link
+  const values = field.value.map((urlInfo: Url, i) => {
+    return <li key={i}>{linkElement(urlInfo, internalOrExternal)}</li>
```
Comment (dgcohen) [inline src/components/BibPage/BibDetail.tsx:112]: This key might also need a prefix/more unique value

File: src/models/BibDetails.ts [hunk 10]
```diff
@@ -0,0 +1,202 @@
...
+      { field: "notes", label: "Notes" },
+      { field: "tableOfContents", label: "Contents" },
+      { field: "shelfMark", label: "Call Number" },
+      { field: "isbn", label: "ISBN" },
+      { field: "issn", label: "ISSN" },
+      { field: "oclc", label: "OCLC" },
+      { field: "lccn", label: "LCCN" },
+    ]
+      .map((fieldMapping) => {
+        let detail
```
Comment (dgcohen) [inline src/models/BibDetails.ts:78]: I believe you don't need to instantiate and assign to a variable, you can just return from within the if statements.

File: src/models/BibDetails.ts [hunk 11]
```diff
@@ -0,0 +1,202 @@
...
+        const filterQueryForSubjectHeading = "/search?filters[subjectLiteral]="
+        // splitSubjectHeadings: ["a", "b", "c"]
+        const splitSubjectHeadings = subject.split(" -- ")
+        return splitSubjectHeadings.map((heading, index) => {
+          const urlWithFilterQuery = `${filterQueryForSubjectHeading}${stackedSubjectHeadings[index]}`
+          const subjectHeadingUrl = {
+            url: urlWithFilterQuery,
+            urlLabel: heading,
+          }
+          return subjectHeadingUrl
```
Comment (dgcohen) [inline src/models/BibDetails.ts:182]: Again you probably don't need to instantiate a variable, you can just return the line above.

File: src/models/BibDetails.ts [hunk 12]
```diff
@@ -0,0 +1,202 @@
...
+    if (!holdings) return []
+    return [
+      { label: "Location", field: "location" },
+      { label: "Format", field: "format" },
+      { label: "Call Number", field: "shelfMark" },
+      { label: "Library Has", field: "holdingStatement" },
+      { label: "Notes", field: "notes" },
+    ]
+      .map((fieldMapping) => {
+        const detail = this.buildHoldingDetail(fieldMapping)
```
Comment (dgcohen) [inline src/models/BibDetails.ts:29]: You can also return this directly without instantiating a variable.

File: src/models/BibDetails.ts [hunk 13]
```diff
@@ -0,0 +1,202 @@
...
+  }
+
+  constructSubjectHeadingsArray(subject: string) {
+    let currentArrayString = ""
+
+    return subject.split(" -- ").map((urlString, index) => {
+      const dashDivided = index !== 0 ? " -- " : ""
+      currentArrayString = `${currentArrayString}${dashDivided}${urlString}`
+
+      return currentArrayString
```
Comment (dgcohen) [inline src/models/BibDetails.ts:199]: Same comment about unneeded variable instantiation.

File: src/utils/bibDetailUtils.ts [hunk 14]
```diff
@@ -0,0 +1,7 @@
+export const displayRtl = (value: string) => {
```
Comment (dgcohen) [inline src/utils/bibDetailUtils.ts:1]: I think these functions could benefit from comments and more explicit naming. I also don't mind them moving to the bibUtils file to cut down on the number of utils files.

File: src/utils/bibModelPreprocessing.ts [hunk 15]
```diff
@@ -0,0 +1,101 @@
...
+   * getNoteType(note)
+   * Construct label for a note by adding the word 'Note'
+   */
+  const getNoteType = (note: Note) => {
+    const type = note.noteType || ""
+    return type.toLowerCase().includes("note") ? type : `${type} (note)`
+  }
+
+  if (!note) {
+    return notesGroupedByNoteType
```
Comment (dgcohen) [inline src/utils/bibModelPreprocessing.ts:17]: You can move this towards the top of the function and return an empty object when it's not set. This makes it more legible/explicit since I think this implies that notesGroupedByNoteType contains something other than an empty object at this point.

File: src/utils/bibModelPreprocessing.ts [hunk 16]
```diff
@@ -0,0 +1,101 @@
...
+
+const getGroupedNotes = (bib: SearchResult) => {
+  const note = bib?.note?.length ? bib.note : null
+  let notesGroupedByNoteType = {}
+
+  /**
+   * getNoteType(note)
+   * Construct label for a note by adding the word 'Note'
+   */
+  const getNoteType = (note: Note) => {
```
Comment (dgcohen) [inline src/utils/bibModelPreprocessing.ts:11]: I think this should be defined outside of this function so it's not instantiated every time it runs.

File: src/utils/bibModelPreprocessing.ts (hunk 15 above)
Comment (dgcohen) [reply src/utils/bibModelPreprocessing.ts:17]: It also might not need to be instantiated at all if you return the note.filter.reduce statement.

File: src/utils/bibModelPreprocessing.ts [hunk 17]
```diff
@@ -0,0 +1,101 @@
...
+  }
+
+  return notesGroupedByNoteType
+}
+
+/**
+ * combineMatching(el1, el2)
+ * Combines properties from matching (i.e. parallel) elements as necessary
+ * Right now, this is only needed to add the 'noteType' in case of parallel notes
+ * @param {string} el1
```
Comment (dgcohen) [inline src/utils/bibModelPreprocessing.ts:42]: Would be good to replace all JS doc annotations with typescript.

File: src/components/BibPage/BibDetail.tsx (hunk 6 above)
Comment (charmingduchess) [reply src/components/BibPage/BibDetail.tsx:51]: sounds good to me!

File: src/models/BibDetails.ts (hunk 11 above)
Comment (charmingduchess) [reply src/models/BibDetails.ts:182]: ah yes thats leftover from debugging

**Comment by charmingduchess** [comment]
@dgcohen thanks for the close eye to the functions!

**Review by dgcohen — APPROVED** [review]
Looks great!
